keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1430 @@
+
+There are cases when the Pod consists of a user container implementing business
+logic and a sidecar providing networking access (Istio), logging (opentelemetry
+collector), or orchestration features. In some cases sidecar containers are
+critical for the user container functioning as they provide a basic
+infrastructure for the user container to run in. In such cases it is beneficial
+to not apply the exponential backoff to the sidecar container restarts and keep
+it at a low constant value.
+
+This is especially true for cases when the sidecar is killed by infrastructure
+(e.g. OOMKill) as it may have happened for reasons independent from the sidecar
+functionality.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+The biggest risk of this proposal is reducing the decay of the _default_
+CrashLoopBackoff: to do so too severely compromises node stability, risking the
+kubelet component to become too slow to respond and the pod lifecycle to
+increase in latency, or worse, causing entire nodes to crash if kubelet takes up
+too much CPU or memory. Since each phase transition for a Pod also has an
+accompanying API request, if the requests become rapid enough due to fast enough
+churn of Pods through CrashLoopBackoff phases, the central API server could
+become unresponsive, effectively taking down an entire cluster.
+
+The same risk exists for the <<[UNRESOLVED]>> per Node <<[/UNRESOLVED]>>
+feature, which, while not default, is by design a more severe reduction in the
+decay behavior. It can abused by <<[UNRESOLVED]>> cluster operators
+<<[/UNRESOLVED]>>, and in the worst case cause nodes to fully saturate with
+<<[UNRESOLVED]>> instantly <<[/UNRESOLVED]>> restarting pods that will never
+recover, risking similar issues as above: taking down nodes or at least
+nontrivially slowing kubelet, or increasing the API requests to store backoff
+state so significantly that the central API server is unresponsive and the
+cluster fails.
+
+During alpha, naturally the first line of defense is that the enhancements, even
+the reduced "default" baseline curve for CrashLoopBackoff, are not usable by
+default and must be opted into. In this specific case they are opted into
+separately as kubelet flags, so clusters will only be affected by each risk if
+the cluster operator enables the new features during the alpha period.
```

Comment:
This makes it sound like there will be explicit configuration knobs to enable this, but I think you're just referring to the off-by-default feature gate?

Reply:
correct, just off-by-default feature gate "during alpha". Clarified in 3763100f1d991a6c4c3a7b4e411805855a34cd84